### PR TITLE
Save post button: avoid extra re-renders when enablng/disabling tooltip

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Fixes https://github.com/WordPress/gutenberg/discussions/44735
Alternative approach to https://github.com/WordPress/gutenberg/pull/56490

This PR makes changes to how the tooltip is rendered for the post save button, thus fixing an issue that was causing loss of keyboard focus on that button.

## Why?

This change is necessary to fix the issue described in https://github.com/WordPress/gutenberg/discussions/44735, where basically the element rendered by `Button` would be removed and re-added to the DOM any time the `Button` component switches from having to show a tooltip to not having to show it, and vice-versa.

The fact that a new HTML element is re-created every time triggers a focus loss that can be very annoying, especially to assistive technology users.

## How?

The approach is similar to what proposed in #56490 — although the fix is applied only for this specific instance of `Button`, instead of applying it directly to the button component (since that [could have broader consequences](https://github.com/WordPress/gutenberg/pull/56490#issuecomment-1825498243)).

The issue is related to the fact that React can't reconciliate the "with tooltip" and "without tooltip" version of the `Button` component without re-creating the whole button every time.

To solve this, I decided to tweak the code so that:
- the `Button` component never renders a tooltip (thanks to removing the `showTooltip` and `shortcut` props, and changing the `label` prop to `aria-label`)
- the `Button` component is always wrapped in a `Tooltip` component. The component receives a different set of props depending on whether a tooltip should be effectively shown.

_Note: this doesn't mean that the tooltip will always be shown, though — the original user-facing behaviour should have been retained by only passing props to the `Tooltip` component when needed._

## Testing Instructions

From https://github.com/WordPress/gutenberg/discussions/44735#discussioncomment-7639716: 

- Edit a draft post and make sure it's saved so that the Save button shows the text 'Saved'.
- Make sure that when focusing or hovering the button, no tooltips are shown
- Make any edit so that the button text changes to 'Save draft'.
- Make sure that when focusing or hovering the button, a tooltip with the text "Save draft ⌘S" is shown
- Select the Save button in the dev tools inspector.
- Right click and enable 'Break on > node removal'.
- Click the Save button.
- Observe the button doesn't gets removed from the DOM, and the keyboard focus stays on the button

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/1083581/aa35002b-60ad-4d28-b8c0-498331147fde

### CODE REVIEW

File: packages/editor/src/components/post-saved-state/index.js
Comment (ciampo) [inline packages/editor/src/components/post-saved-state/index.js:1]: Reviewing the diff with the "Hide whitespace" option enabled can be quite helpful here

File: packages/editor/src/components/post-saved-state/index.js
Comment (ciampo) [inline packages/editor/src/components/post-saved-state/index.js:165]: the `label` prop has two functions in button:
- it serves as an aria-label, if the aria-label prop isn't explicitly set
- it's used as tooltip text

Swapping `label` for `aria-label` allows us to keep the button correctly labelled, without triggering the internal `Button`'s tooltip.

The same string of text is also being passed explicitly to the `Tooltip` component via the `text` prop.

File: packages/editor/src/components/post-saved-state/index.js
Comment (ciampo) [inline packages/editor/src/components/post-saved-state/index.js:154]: This prop is now passed to the `Tooltip` component instead

**Comment by ciampo** [comment]
I think that we could merge this PR anyway, while we discuss whether the approach in #56490 is viable.

File: packages/editor/src/components/post-saved-state/index.js
Comment (afercia) [inline packages/editor/src/components/post-saved-state/index.js]: Spotted a couplle typos: `only is the button` should be `only if the button`, ` always remove and` should be `always removed and`.

File: packages/editor/src/components/post-saved-state/index.js
Comment (afercia) [inline packages/editor/src/components/post-saved-state/index.js:153]: I think this comment block should be removed now.

File: packages/editor/src/components/post-saved-state/index.js
Comment (afercia) [inline packages/editor/src/components/post-saved-state/index.js]: I think this comment block should be removed now.

**Comment by afercia** [comment]
Thanks @ciampo, I left a couple minor comments.
To my understanding, this would only fix the Save/Saved button. It doesn't actually fix the genereal case reported in #44735. 
I have no objections to fix the Saved button specific case but #44735 should not be closed.

File: packages/editor/src/components/post-saved-state/index.js
Comment (ciampo) [reply packages/editor/src/components/post-saved-state/index.js]: https://github.com/WordPress/gutenberg/pull/56502/commits/7bce2ea963ac5c37258ae7f5995041905077afba

File: packages/editor/src/components/post-saved-state/index.js
Comment (ciampo) [reply packages/editor/src/components/post-saved-state/index.js]: https://github.com/WordPress/gutenberg/pull/56502/commits/033828f8c9189ade962303058283d8bdc102f71d

**Comment by ciampo** [comment]
> To my understanding, this would only fix the Save/Saved button. It doesn't actually fix the genereal case reported in https://github.com/WordPress/gutenberg/discussions/44735.

That's correct. A more general fix is being discussed in #56490, although I'm not sure how easy it's going to be to find a solid solution that doesn't introduce regressions.

This PR offers a fix specifically to the Save button, and shows an alternative workaround to setting the `showTooltip` prop on the `Button` component.

> I have no objections to fix the Saved button specific case but https://github.com/WordPress/gutenberg/discussions/44735 should not be closed.

Since the button is currently working as expected thanks to having the `showTooltip` prop, the fix from this PR isn't urgent — so I'm also OK if we don't merge it.

Regardless, I think that it would be better for #44735 to become an actual GH issue, rather than a discussion,

File: packages/editor/src/components/post-saved-state/index.js
Comment (tyxla) [inline packages/editor/src/components/post-saved-state/index.js]: Just noticed we can shorten this to:

```suggestion
				onClick={ isDisabled ? undefined : savePost }
```

**Review by tyxla — APPROVED** [review]
Looks good code-wise and tests well 👍 

I think we can merge this one and keep the https://github.com/WordPress/gutenberg/discussions/44735 discussion going separately.

**Comment by afercia** [comment]
> Regardless, I think that it would be better for https://github.com/WordPress/gutenberg/discussions/44735 to become an actual GH issue, rather than a discussion,

I forgot https://github.com/WordPress/gutenberg/discussions/44735 is a discussion and not a GitHub issue 🤦 
I believe people with triage permission can create an issue from a discussion.

**Comment by tyxla** [comment]
> I believe people with triage permission can create an issue from a discussion.

Yes, there should be a "Create issue from discussion" link in the sidebar of the discussion page:

sidebar

**Comment by ciampo** [comment]
> Looks good code-wise and tests well 👍
> 
> I think we can merge this one and keep the #44735 discussion going separately.

Sounds good!

I created https://github.com/WordPress/gutenberg/issues/56522 that can be used instead of the original discussion.

File: packages/editor/src/components/post-saved-state/index.js
Comment (ramonjd) [inline packages/editor/src/components/post-saved-state/index.js:169]: It looks like the [JS unit tests](https://github.com/WordPress/gutenberg/actions/runs/6985585167/job/19010003602) were failing on this branch before it was merged due to this line?

This test checks that the event is **not** passed to the store action `savePost`: 

https://github.com/WordPress/gutenberg/blob/5d86605316cc339cb8f15264a3fdf395f935618f/packages/editor/src/components/post-saved-state/test/index.js#L88-L88

I don't know how important that is, but I guess we shouldn't be dispatching actions with such values 🤷🏻 

This fixes it (and I think is unrelated to the tooltip/rerender issue)

```diff
diff --git a/packages/editor/src/components/post-saved-state/index.js b/packages/editor/src/components/post-saved-state/index.js
index c3bbad2268..7b2c19d6ea 100644
--- a/packages/editor/src/components/post-saved-state/index.js
+++ b/packages/editor/src/components/post-saved-state/index.js
@@ -166,7 +166,7 @@ export default function PostSavedState( {
 						  } )
 						: undefined
 				}
-				onClick={ isDisabled ? undefined : savePost }
+				onClick={ isDisabled ? undefined : () => savePost() }
 				variant="tertiary"
 				icon={ isLargeViewport ? undefined : cloudUpload }
 				// Make sure the aria-label has always a value, as the default `text` is undefined on small screens.
```

I'll throw a PR up so that trunk tests are fixed anyway.

If we want to avoid creating a new function ref every render then I guess we could go the extra step and throw a `useCallback` in there?

```js
	const onClickHandler = useCallback(
		() => ( isDisabled ? undefined : savePost() ),
		[ isDisabled ]
```

_Edit_: 

- https://github.com/WordPress/gutenberg/pull/56529

**Pull request merged**
